### PR TITLE
Reduced allocations in Mailboxes.GetMailboxType

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
## Before
```
TotalBytesAllocated: Max: 1 502 856,00 bytes, Average: 1 457 532,92 bytes, Min: 1 453 712,00 bytes, StdDev: 13 617,96 bytes
TotalBytesAllocated: Max / s: 405 371,67 bytes, Average / s: 392 464,67 bytes, Min / s: 371 907,99 bytes, StdDev / s: 11 683,17 bytes

TotalCollections [Gen0]: Max: 1 050,00 collections, Average: 1 050,00 collections, Min: 1 050,00 collections, StdDev: 0,00 collections
TotalCollections [Gen0]: Max / s: 292,79 collections, Average / s: 282,79 collections, Min / s: 259,84 collections, StdDev / s: 9,94 collections

TotalCollections [Gen1]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen1]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

TotalCollections [Gen2]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen2]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

[Counter] ActorCreateThroughput: Max: 1 000 000,00 operations, Average: 1 000 000,00 operations, Min: 1 000 000,00 operations, StdDev: 0,00 operations
[Counter] ActorCreateThroughput: Max / s: 278 851,27 operations, Average / s: 269 323,11 operations, Min / s: 247 467,48 operations, StdDev / s: 9 470,60 operations
```
## After
```
TotalBytesAllocated: Max: 2 860 784,00 bytes, Average: 2 821 197,54 bytes, Min: 2 787 816,00 bytes, StdDev: 37 255,75 bytes
TotalBytesAllocated: Max / s: 887 349,58 bytes, Average / s: 858 735,53 bytes, Min / s: 779 925,67 bytes, StdDev / s: 27 389,70 bytes

TotalCollections [Gen0]: Max: 986,00 collections, Average: 986,00 collections, Min: 986,00 collections, StdDev: 0,00 collections
TotalCollections [Gen0]: Max / s: 313,84 collections, Average / s: 300,25 collections, Min / s: 268,90 collections, StdDev / s: 12,51 collections

TotalCollections [Gen1]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen1]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

TotalCollections [Gen2]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen2]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

[Counter] ActorCreateThroughput: Max: 1 000 000,00 operations, Average: 1 000 000,00 operations, Min: 1 000 000,00 operations, StdDev: 0,00 operations
[Counter] ActorCreateThroughput: Max / s: 318 294,69 operations, Average / s: 304 514,73 operations, Min / s: 272 722,64 operations, StdDev / s: 12 686,10 operations
```

### CODE REVIEW

File: src/core/Akka/Actor/Deployer.cs [hunk 1]
```diff
@@ -54,10 +54,20 @@ public Deployer(Settings settings)
...
         {
-            if (path.Elements.Head() != "user" || path.Elements.Count() < 2)
+            var rawElements = path.Elements;
+            if (rawElements[0] != "user" || rawElements.Count < 2)
+            {
                 return Deploy.None;
+            }
+
+            var elements = new List<string>(rawElements.Count);
+            for (int i = 0; i < rawElements.Count; i++)
```
Comment (Rattenkrieg) [inline src/core/Akka/Actor/Deployer.cs]: may be it will be cleaner to start from `i = 1`

**Comment by Horusiath** [comment]
Before: 

> TotalBytesAllocated: Max: 1 502 856,00 bytes, Average: 1 457 532,92 bytes, Min: 1 453 712,00 bytes, StdDev: 13 617,96 bytes
TotalBytesAllocated: Max / s: 405 371,67 bytes, Average / s: 392 464,67 bytes, Min / s: 371 907,99 bytes, StdDev / s: 11 683,17 bytes

After:

>TotalBytesAllocated: Max: 2 860 784,00 bytes, Average: 2 821 197,54 bytes, Min: 2 787 816,00 bytes, StdDev: 37 255,75 bytes
TotalBytesAllocated: Max / s: 887 349,58 bytes, Average / s: 858 735,53 bytes, Min / s: 779 925,67 bytes, StdDev / s: 27 389,70 bytes

Why bytes allocated after this change have increased so much?

**Comment by alexvaluyskiy** [comment]
@Horusiath Have no idea. I make only one allocation `var elements = new List<string>(rawElements.Count);`, and removed many allocation related with LINQ

**Comment by rogeralsing** [comment]
>Why bytes allocated after this change have increased so much?

My guess would be the extra copy to a list instead of passing the enumerable to Lookup(...)

These bits are the major bottleneck for Akka.NET in the SkynetBenchmark.
Enabling Server GC for the spawn test drops the total time from 14 sec to 6!
Havent verified with this PR. but for Akka.NET to be closer to all other libs in terms of actor weight/cost, this needs fixing.

**Comment by alexvaluyskiy** [comment]
@rogeralsing But previouly we were calling ActorPath.Elements setter three times (at least 9 allocations)
https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka/Actor/ActorPath.cs#L200

And allocated a new list - `var elements = path.Elements.Drop(1);`
https://github.com/akkadotnet/akka.net/blob/2e59b2b204dff7b9dcbe93663077b0230944f18b/src/core/Akka/Util/Internal/Extensions.cs#L37

**Comment by alexvaluyskiy** [comment]
@Horusiath I've updated the PR, now
```
TotalBytesAllocated: Max: 626 408,00 bytes, Average: 596 929,85 bytes, Min: 580 312,00 bytes, StdDev: 11 820,18 bytes
TotalBytesAllocated: Max / s: 261 336,04 bytes, Average / s: 251 956,03 bytes, Min / s: 243 995,22 bytes, StdDev / s: 4 748,45 bytes

TotalCollections [Gen0]: Max: 946,00 collections, Average: 946,00 collections, Min: 946,00 collections, StdDev: 0,00 collections
TotalCollections [Gen0]: Max / s: 403,68 collections, Average / s: 399,31 collections, Min / s: 394,22 collections, StdDev / s: 2,98 collections

TotalCollections [Gen1]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen1]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

TotalCollections [Gen2]: Max: 0,00 collections, Average: 0,00 collections, Min: 0,00 collections, StdDev: 0,00 collections
TotalCollections [Gen2]: Max / s: 0,00 collections, Average / s: 0,00 collections, Min / s: 0,00 collections, StdDev / s: 0,00 collections

[Counter] GetMailboxTypeFootprint: Max: 1 000 000,00 operations, Average: 1 000 000,00 operations, Min: 1 000 000,00 operations, StdDev: 0,00 operations
[Counter] GetMailboxTypeFootprint: Max / s: 426 727,32 operations, Average / s: 422 103,75 operations, Min / s: 416 727,97 operations, StdDev / s: 3 151,54 operations
```

**Comment by Horusiath** [comment]
I don't know how @Rattenkrieg feels about that, but for me it's good to go once it'll be rebased.

**Comment by Aaronontheweb** [comment]
@alexvaluyskiy updated benchmark results after #2549 


### Totals
          Metric |           Units |             Max |         Average |             Min |          StdDev |
---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
TotalBytesAllocated |           bytes |      746,336.00 |      735,860.92 |      734,608.00 |        3,169.66 |
TotalCollections [Gen0] |     collections |           80.00 |           80.00 |           80.00 |            0.00 |
TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
[Counter] GetMailboxTypeFootprint |      operations |    1,000,000.00 |    1,000,000.00 |    1,000,000.00 |            0.00 |

### Per-second Totals
          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
TotalBytesAllocated |           bytes |    1,285,533.56 |    1,248,297.36 |    1,094,987.61 |       49,525.71 |
TotalCollections [Gen0] |     collections |          140.00 |          135.73 |          117.37 |            5.87 |
TotalCollections [Gen1] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
TotalCollections [Gen2] |     collections |            0.00 |            0.00 |            0.00 |            0.00 |
[Counter] GetMailboxTypeFootprint |      operations |    1,749,958.57 |    1,696,656.00 |    1,467,151.00 |       73,435.34 |

**Comment by alexvaluyskiy** [comment]
Very good

**Comment by Aaronontheweb** [comment]
@alexvaluyskiy I'm feeling most good about the GC numbers... Down to 80 GC 0 collections from 1,050 at the start.

**Comment by Aaronontheweb** [comment]
Also, total actor create throughput is up another 10k / s with these changes from what we recorded on #2549 

### Totals
          Metric |           Units |             Max |         Average |             Min |          StdDev |
---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
TotalBytesAllocated |           bytes |   19,299,648.00 |   18,139,634.46 |   16,175,960.00 |      902,130.54 |
[Counter] ActorCreateThroughput |      operations |       10,000.00 |       10,000.00 |       10,000.00 |            0.00 |

### Per-second Totals
          Metric |       Units / s |         Max / s |     Average / s |         Min / s |      StdDev / s |
---------------- |---------------- |---------------- |---------------- |---------------- |---------------- |
TotalBytesAllocated |           bytes |  161,833,945.51 |  139,934,829.45 |  111,091,530.55 |   16,057,589.65 |
[Counter] ActorCreateThroughput |      operations |       84,379.85 |       76,949.91 |       68,676.93 |        5,718.04 |

**Pull request merged**